playerctl/playerctl-cli.c

```diff
@@ -16,279 +16,359 @@
+      G_BUS_TYPE_SESSION,
+      G_DBUS_PROXY_FLAGS_NONE,
+      NULL,
+      "org.freedesktop.DBus",
+      "/org/freedesktop/DBus",
+      "org.freedesktop.DBus",
+      NULL,
+      &tmp_error);
 
-  context = g_option_context_new("- Controller for MPRIS players");
-  g_option_context_add_main_entries(context, entries, NULL);
-  g_option_context_set_description(context, description);
-  g_option_context_set_summary(context, summary);
+  if (tmp_error != NULL) {
+    g_propagate_error(err, tmp_error);
+    return NULL;
+  }
 
-  succes = g_option_context_parse(context, &argc, &argv, &error);
-  g_option_context_free (context);
+  GVariant *reply = g_dbus_proxy_call_sync(proxy,
+      "ListNames",
+      NULL,
+      G_DBUS_CALL_FLAGS_NONE,
+      -1,
+      NULL,
+      &tmp_error);
 
-  if (!succes) {
-    g_printerr("Option parsing failed: %s\n", error->message);
-    return 1;
+  if (tmp_error != NULL) {
+    g_propagate_error(err, tmp_error);
+    g_object_unref(proxy);
+    return NULL;
   }
 
-  if (version_opt) {
-    g_print("v%s\n", PLAYERCTL_VERSION_S);
-    return 0;
+  GString *names_str = g_string_new("");
+  GVariant *reply_child = g_variant_get_child_value(reply, 0);
+  gsize reply_count;
+  const gchar **names = g_variant_get_strv(reply_child, &reply_count);
+
+  for (int i = 0; i < reply_count; i += 1) {
+    if (g_str_has_prefix(names[i], "org.mpris.MediaPlayer2")) {
+      g_string_append_printf(names_str, "%s\n", names[i] + 23);
+    }
   }
 
-  if (list_all_opt) {
-    gchar *player_names = list_player_names(&error);
+  g_object_unref(proxy);
+  g_variant_unref(reply);
+  g_variant_unref(reply_child);
+  g_free(names);
 
-    if (error != NULL) {
-      g_printerr("Could not list players: %s\n", error->message);
-      g_clear_error(&error);
-      return 1;
-    }
+  return g_string_free(names_str, FALSE);
+}
 
-    if (player_names[0] == '\0')
-      g_printerr("%s\n", "No players were found");
-    else
-      g_print("%s", player_names);
-    g_free(player_names);
+static gboolean list_all_players_and_exit (const gchar *name, const gchar *value, PlayerctlPlayer *player, GError **error)
+{
+  GError *tmp_error = NULL;
+  gchar *player_names = list_player_names(&tmp_error);
 
-    return 0;
+  if (tmp_error) {
+    g_propagate_error(error, tmp_error);
+    return FALSE;
   }
 
-  if (command == NULL) {
-    g_print("%s", g_option_context_get_help(context, TRUE, NULL));
-    return 0;
-  }
+  if (player_names[0] == '\0')
+    g_printerr("No players were found");
+  else
+    g_print("%s", player_names);
+  g_free(player_names);
 
-  PlayerctlPlayer *player = playerctl_player_new(player_name, &error);
+  exit(0);
+  return TRUE;
+}
 
-  if (error != NULL) {
-    g_printerr("Connection to player failed: %s\n", error->message);
-    g_clear_error(&error);
-    return 1;
-  }
+static gboolean print_version_and_exit (const gchar *name, const gchar *value, PlayerctlPlayer *player, GError **error)
```

Comment:
I went ahead and restored this to a global variable. Let me know if you want me to solve this differently.